docs/design-documents/20220408-loading-pipeline-configuration-files.md

```diff
@@ -0,0 +1,270 @@
+file to be set for Conduit to read.
+
+```go
+// Config holds all configurable values for Conduit.
+type Config struct {
+	// File holds a Path to a configuration file that Conduit should load.
+	File struct {
+		Path string
+	}
+
+	DB struct {
+		Type   string
+		Badger struct {
+			Path string
+		}
+		Postgres struct {
+			ConnectionString string
+			Table            string
+		}
+	}
+
+	HTTP struct {
+		Address string
+	}
+	GRPC struct {
+		Address string
+	}
+
+	Log struct {
+		Level  string
+		Format string
+	}
+}
+```
+
+The above configuration is parsed at start time and provies a reference to our 
+pipeline configuration file. After parsing that file, we inject environment 
+variables into their corresponding templated locations, if any.
+
+```go
+// DB defines the interface for a key-value store.
+type DB interface {
+	// NewTransaction starts and returns a new transaction.
+	NewTransaction(ctx context.Context, update bool) (Transaction, context.Context, error)
+	// Close should flush any cached writes, release all resources and close the
+	// store. After Close is called other methods should return an error.
+	Close() error
+
+	Set(ctx context.Context, key string, value []byte) error
+	Get(ctx context.Context, key string) ([]byte, error)
+	GetKeys(ctx context.Context, prefix string) ([]string, error)
+}
+```
+
+### Config files as a database implementation
+
+`DB` is an interface for storing key-value pairs. Our `ConfigFileAsDatabase` 
+struct will store a reference to a parsed configuration file and 
+environment and maps that to a `Store` for each `Service`.
+
+A key part about `ConfigFileAsADatabase` is that it stubs out mutable
+operations, including transaction handling, enforcing our immutability
+during operation.
+
+We pass that `ConfigFileAsADatabase` to each new service at start time.
+Our `ConfigFileAsDatabase` will be bound with an in-memory database and passed
+to each new service when Conduit is starting.
+
+```go
+// NewService initializes and returns a pipeline Service.
+func NewService(logger log.CtxLogger, db database.DB) *Service {
+	return &Service{
+		logger:        logger.WithComponent("pipeline.Service"),
+		store:         NewStore(db),
+		instances:     make(map[string]*Instance),
+		instanceNames: make(map[string]bool),
+	}
+}
+
+svc := NewService(logger, &ConfigAsDatabase{})
+```
+
+### Hydration 
+
+Once Conduit's services are started, they will be empty. We must parse our 
+config file into a full pipeline by hydrating it with other necessary 
+information including environment information, connector configurations, 
+and processors.
+
+This step will be responsible for populating the service's individual `Stores` 
+with the necessary resources before `Init` is called during start.
+
+This process can fail, in which a pipeline reports `StatusDegraded`.
+Because we aren't dictating restart behavior in configuration files, this
+should cause Conduit to fail and report the pipeline errors it received.
```

Comment:
Are we saying that Conduit should stop running if a pipeline defined in a config file encounters an error? If so, that would mean any other pipeline running on that instance would be stopped as well 😕  If a user wanted resilience, they would have to run one Conduit instance per pipeline, I'm not convinced it's the right thing to do.

Reply:
I should clear this up, but no, that is not what was intended. Conduit's current behavior of failing the pipeline is what I was meaning to describe here, no changes to that are necessary in my opinion.